### ISSUE TITLE
ALX:  a function that searches for an integer.

### ISSUE THREAD

**Issue by ahmed149** [ahmed149/learning-programing-in-C#69]
#include "function_pointers.h"

/**
  * int_index - ...
  * @array: ...
  * @size: ...
  * @cmp: ...
  *
  * Return: ...
  */
int int_index(int *array, int size, int (*cmp)(int))
{
	int i = 0;

	if (size > 0)
	{
		if (array != NULL && cmp != NULL)
		{
			while (i < size)
			{
				if (cmp(array[i]))
					return (i);

				i++;
			}
		}
	}

	return (-1);
}

**Comment by ahmed149** [comment]
Certainly! This C function, `int_index`, is designed to search for an integer element in an array and return the index at which the first element that matches a specific condition (defined by a function `cmp`) is found. Here's an explanation of each part of the function:

1. `#include "function_pointers.h"`:
   - This line includes a header file called "function_pointers.h." It's common to include header files at the beginning of C files to declare functions and data structures used in the program.

2. `int int_index(int *array, int size, int (*cmp)(int))`:
   - This is the function declaration for `int_index`. It takes three parameters:
     - `array`: A pointer to an integer array.
     - `size`: The size (number of elements) of the array.
     - `cmp`: A function pointer named `cmp` that takes an integer as an argument and returns an integer. This function pointer is used to specify the condition for comparison.

3. `if (size > 0)`:
   - This condition checks whether the `size` is greater than 0. It ensures that the array has at least one element to search.

4. `if (array != NULL && cmp != NULL)`:
   - Inside the previous condition, it checks whether the `array` pointer is not NULL (i.e., it points to a valid array) and whether the `cmp` pointer is not NULL (i.e., it points to a valid comparison function).

5. `while (i < size)`:
   - Inside the inner conditions, it enters a `while` loop that iterates from 0 to `size - 1`. It uses the `i` variable to keep track of the current index.

6. `if (cmp(array[i]))`:
   - Inside the loop, it calls the function pointed to by `cmp` with `array[i]` as its argument. If the result of this function call is non-zero (true), it means that the condition specified by `cmp` is met for the current element, and it returns the index `i` immediately.

7. `return (-1);`:
   - If no matching element is found in the array, or if the input conditions (size, array, cmp) are not met, the function returns -1 to indicate that no element meets the specified condition.

In summary, the `int_index` function is designed to search for an element in an integer array based on a specified condition (defined by the `cmp` function). It returns the index of the first element that satisfies the condition or -1 if no such element is found or if the input conditions are not met.

Here's a simplified example of how you might use this function:

```c
#include <stdio.h>

int is_even(int x)
{
    return (x % 2 == 0);
}

int main()
{
    int arr[] = {1, 3, 5, 8, 10, 12};
    int size = sizeof(arr) / sizeof(arr[0]);

    // Use int_index to find the index of the first even number in the array
    int index = int_index(arr, size, is_even);

    if (index != -1)
    {
        printf("First even number found at index %d\n", index);
    }
    else
    {
        printf("No even number found in the array.\n");
    }

    return 0;
}
```

In this example, the `int_index` function is used to find the index of the first even number in the `arr` array using the `is_even` function as the comparison condition.

**Comment by ahmed149** [comment]
Certainly! The line `int size = sizeof(arr) / sizeof(arr[0]);` is used to calculate the number of elements in an array named `arr` and store it in the integer variable `size`. Let me break down this line step by step:

1. `sizeof(arr)`:
   - `sizeof` is an operator in C that returns the size in bytes of its operand. In this case, the operand is `arr`, which is an array.

2. `sizeof(arr[0])`:
   - This part calculates the size in bytes of a single element of the `arr` array. `arr[0]` refers to the first element of the array. By dividing the total size of the array by the size of a single element, you can determine how many elements are in the array.

3. `sizeof(arr) / sizeof(arr[0])`:
   - This division calculates the number of elements in the array `arr`. It divides the total size of the array (in bytes) by the size of a single element (in bytes). This division results in the count of elements in the array, which is an integer.

4. `int size = ...;`:
   - Finally, the result of the division is assigned to the integer variable `size`. So, `size` now contains the number of elements in the `arr` array.

In summary, this line of code is a common technique in C to find the size or number of elements in an array dynamically. It's useful because it adapts to changes in the array size automatically, making your code more flexible and less error-prone.
